// primeng/esm2022/inputmask/style/inputmaskstyle.mjs
import { Injectable } from '@angular/core';
import { BaseStyle } from 'primeng/base';
import * as i0 from "@angular/core";
const theme = ({ dt }) => `
p-inputmask {
    position: relative;
}

.p-inputmask-clear-icon {
    position: absolute;
    top: 50%;
    margin-top: -0.5rem;
    cursor: pointer;
    inset-inline-end: ${dt('form.field.padding.x')};
    color: ${dt('form.field.icon.color')};
}

p-inputmask.ng-invalid.ng-dirty > .p-inputtext {
    border-color: ${dt('inputtext.invalid.border.color')};
}

p-inputmask.ng-invalid.ng-dirty > .p-inputtext:enabled:focus {
    border-color: ${dt('inputtext.focus.border.color')};
}

p-inputmask.ng-invalid.ng-dirty > .p-inputtext::placeholder {
    color: ${dt('inputtext.invalid.placeholder.color')};
}
`;
const classes = {
    root: ({ instance }) => ({
        'p-inputmask': true,
        'p-filled': instance.variant ? instance.variant === 'filled' : instance.config.inputStyle() === 'filled'
    })
};
export class InputMaskStyle extends BaseStyle {
    name = 'inputmask';
    theme = theme;
    classes = classes;
    static ɵfac = /*@__PURE__*/ (() => { let ɵInputMaskStyle_BaseFactory; return function InputMaskStyle_Factory(__ngFactoryType__) { return (ɵInputMaskStyle_BaseFactory || (ɵInputMaskStyle_BaseFactory = i0.ɵɵgetInheritedFactory(InputMaskStyle)))(__ngFactoryType__ || InputMaskStyle); }; })();
    static ɵprov = /*@__PURE__*/ i0.ɵɵdefineInjectable({ token: InputMaskStyle, factory: InputMaskStyle.ɵfac });
}
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(InputMaskStyle, [{
        type: Injectable
    }], null, null); })();
/**
 *
 * InputMask component is used to enter input in a certain format such as numeric, date, currency, email and phone.
 *
 * [Live Demo](https://www.primeng.org/inputmask/)
 *
 * @module inputmaskstyle
 *
 */
export var InputMaskClasses;
(function (InputMaskClasses) {
    /**
     * Class name of the root element
     */
    InputMaskClasses["root"] = "p-inputmask";
})(InputMaskClasses || (InputMaskClasses = {}));
